File: backend/transcendence/game/game_logic/Ball.py
from .logic_utils import get_position_percentage
from datetime import datetime
from .const_vars import *
import random
import math

class Ball:
	def __init__(self):
		self.screen_height = SCREEN_HEIGHT
		self.screen_width = SCREEN_WIDTH
		self.x_start = self.screen_width / 2
		self.y_start = self.screen_height / 2
		self.x = self.x_start
		self.y = self.y_start
		self.last_time = 0
		self.__set_start_angle()
		self.left_wall_limit = BALL_RADIUS
		self.right_wall_limit = self.screen_width - BALL_RADIUS
		self.top_wall_limit = BALL_RADIUS
		self.bottom_wall_limit = self.screen_height - BALL_RADIUS

	def get_position(self):
		positions = {
			"x": get_position_percentage(self.x, SCREEN_WIDTH),
			"y": get_position_percentage(self.y, SCREEN_HEIGHT)
		}
		return positions

	def update_position(self, left_paddle, right_paddle):
		radius = self.__get_moved_distance()
		x = self.x + radius * self.trig_values["cos_value"]
		y = self.y + radius * self.trig_values["sin_value"]

		colision_coords = self.__get_colision_point(y)
		paddle_colisions_coords = self.__get_paddle_colisions_coords(x, y, left_paddle, right_paddle)

		if colision_coords:
			self.x = colision_coords['x']
			self.y = colision_coords['y']
			self.__set_new_reflection_angle()
		elif paddle_colisions_coords:
			self.x = paddle_colisions_coords['x']
			self.y = paddle_colisions_coords['y']
			self.__set_new_paddle_angle(paddle_colisions_coords['hit_paddle_percentage'])
		else:
			self.x = x
			self.y = y

	def goal_detection(self):
		player_1_goal = self.x >= self.right_wall_limit
		player_2_goal = self.x <= self.left_wall_limit 
		if player_1_goal or player_2_goal:
			goal_info = {
				"player_1": player_1_goal,
				"player_2": player_2_goal
			}
			self.x = self.x_start
			self.y = self.y_start
			self.__set_start_angle()
			return goal_info
		return None

	def set_end_game_position(self):
		self.x = self.x_start
		self.y = self.y_start

	def __get_moved_distance(self):
		if not self.last_time:
			self.last_time = datetime.now()
			value = 1
		else:
			current_time = datetime.now()
			value = (current_time.timestamp() - self.last_time.timestamp()) * 1000
			self.last_time = current_time
		radius = value * BALL_SPEED
		return radius

	def __set_trig_values(self, angle_rad):
		self.trig_values = {
			"sin_value": math.sin(angle_rad) * -1,
			"cos_value": math.cos(angle_rad)
		}

	def __get_colision_point(self, y):
		if y > self.bottom_wall_limit:
			col_y = self.bottom_wall_limit
		elif y < self.top_wall_limit:
			col_y = BALL_RADIUS
		else:
			return None
		m = math.tan(self.angle_rad)
		b = self.y - m * self.x
		col_x = (col_y - b) / m
		return {"x": col_x, "y": col_y}

	def __set_new_reflection_angle(self):
		self.__set_angle(360 - self.angle_deg)

	def __set_start_angle(self):
		self.__set_angle(self.__get_start_angle())

	def __set_angle(self, angle_deg):
		if angle_deg < 0:
			angle_deg += 360
		elif angle_deg > 360:
			angle_deg -= 360
		self.angle_deg = angle_deg
		self.angle_rad = math.radians(angle_deg)
		self.__set_trig_values(self.angle_rad)

	def __get_paddle_colisions_coords(self, x, y, left_paddle, right_paddle):
		if self.angle_deg > 90 and self.angle_deg < 270:
			new_coords = left_paddle.get_colision_point(self.x, self.y, x, y, BALL_RADIUS)
		else:
			new_coords = right_paddle.get_colision_point(self.x, self.y, x, y, BALL_RADIUS)
		return new_coords

	def __set_new_paddle_angle(self, hit_percentage):
		if self.angle_deg > 90 and self.angle_deg < 270:
			result_angle = 420 + (hit_percentage/100) * (300 - 420)
		else:
			result_angle = 120 + (hit_percentage/100) * (240 - 120)
		self.__set_angle(result_angle)

	def __get_start_angle(self):
		chosen_interval = random.choice(ANGLES_INTERVALS)
		start_angle = 180
		while start_angle == 180 or start_angle == 360:
			start_angle = random.randint(chosen_interval[0], chosen_interval[1])
		return start_angle

File: srcs/frontend/app/game/Ball.js
import * as const_vars from "./const_vars.js";
import get_position_percentage from "./logicUtils.js"

const BALL_RADIUS = 5;

export class Ball {
	constructor() {
		this.screen_height = const_vars.SCREEN_HEIGHT;
		this.screen_width = const_vars.SCREEN_WIDTH;
		this.x_start = this.screen_width / 2;
		this.y_start = this.screen_height / 2;
		this.x = this.x_start;
		this.y = this.y_start;
		this.last_time = 0;
		this.#setStartAngle();
		this.left_wall_limit = const_vars.BALL_RADIUS;
		this.right_wall_limit = this.screen_width - const_vars.BALL_RADIUS;
		this.top_wall_limit = const_vars.BALL_RADIUS;
		this.bottom_wall_limit = this.screen_height - const_vars.BALL_RADIUS;
	}

	getPosition() {
		return {
			x: get_position_percentage(this.x, const_vars.SCREEN_WIDTH),
			y: get_position_percentage(this.y, const_vars.SCREEN_HEIGHT)
		};
	}

	updatePosition(leftPaddle, rightPaddle) {
		const radius = this.#getMovedDistance();
		const x = this.x + radius * this.trig_values.cos_value;
		const y = this.y + radius * this.trig_values.sin_value;
		const colisionCoords = this.#getColisionPoint(y);
		const paddleColisionsCoords = this.#getPaddleColisionsCoords(x, y, leftPaddle, rightPaddle);

		if (colisionCoords) {
			this.x = colisionCoords.x;
			this.y = colisionCoords.y;
			this.#setNewReflectionAngle();
		}
		else if (paddleColisionsCoords) {
			this.x = paddleColisionsCoords.x;
			this.y = paddleColisionsCoords.y;
			this.#setNewPaddleAngle(paddleColisionsCoords.hit_paddle_percentage);
		}
		else {
			this.x = x;
			this.y = y;
		}
	}

	goalDetection() {
		const player1Goal = this.x >= this.right_wall_limit;
		const player2Goal = this.x <= this.left_wall_limit;
		if (player1Goal || player2Goal) {
			const goalInfo = { player1: player1Goal, player2: player2Goal };
			this.x = this.x_start;
			this.y = this.y_start;
			this.#setStartAngle();
			return goalInfo;
		}
		return null;
	}

	setEndGamePosition() {
		this.x = this.x_start;
		this.y = this.y_start;
	}

	#getMovedDistance() {
		let value;
		if (!this.last_time) {
			this.last_time = Date.now();
			value = 1;
		}
		else {
			const current_time = Date.now();
			value = (current_time - this.last_time);
			this.last_time = current_time;
		}
		return value * const_vars.BALL_SPEED;
	}

	#setTrigValues(angleRad) {
		this.trig_values = {
			sin_value: -Math.sin(angleRad),
			cos_value: Math.cos(angleRad),
		};
	}

	#getColisionPoint(y) {
		let col_y;
		if (y > this.bottom_wall_limit)
			col_y = this.bottom_wall_limit;
		else if (y < this.top_wall_limit)
			col_y = const_vars.BALL_RADIUS;
		else
			return null;
		const m = Math.tan(this.angle_rad);
		const b = this.y - m * this.x;
		const col_x = (col_y - b) / m;
		return { x: col_x, y: col_y };
	}

	#setNewReflectionAngle() {
		this.#setAngle(360 - this.angle_deg);
	}

	#setStartAngle() {
		const chosenInterval = const_vars.ANGLES_INTERVALS[Math.floor(Math.random() * const_vars.ANGLES_INTERVALS.length)];
		this.#setAngle(Math.floor(Math.random() * (chosenInterval[1] - chosenInterval[0] + 1)) + chosenInterval[0]);
	}

	#setAngle(angleDeg) {
		if (angleDeg < 0)
			angleDeg += 360;
		else if (angleDeg > 360)
			angleDeg -= 360;
		this.angle_deg = angleDeg;
		this.angle_rad = this.angle_deg * (Math.PI / 180);
		this.#setTrigValues(this.angle_rad);
	}

	#getPaddleColisionsCoords(x, y, leftPaddle, rightPaddle) {
		let newCoords;
		if (this.angle_deg > 90 && this.angle_deg < 270)
			newCoords = leftPaddle.getColisionPoint(this.x, this.y, x, y, const_vars.BALL_RADIUS);
		else
			newCoords = rightPaddle.getColisionPoint(this.x, this.y, x, y, const_vars.BALL_RADIUS);
		return newCoords;
	}

	#setNewPaddleAngle(hitPercentage) {
		let resultAngle;
		if (this.angle_deg > 90 && this.angle_deg < 270)
			resultAngle = 420 + (hitPercentage / 100) * (300 - 420);
		else
			resultAngle = 120 + (hitPercentage / 100) * (240 - 120);
		this.#setAngle(resultAngle);
	}
}
